lightning/src/ln/channelmanager.rs

```diff
@@ -3056,6 +3087,23 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 		let (pending_forward_info, mut channel_state_lock) = self.decode_update_add_htlc_onion(msg);
 		let channel_state = &mut *channel_state_lock;
 
+		// If we're receiving a keysend payment, then no payment will have been added to
+		// `self.pending_inbound_payments` in preparation to receive said payment. Thus, if a
+		// keysend_preimage is present, add the pending inbound payment now.
+		match pending_forward_info {
+			PendingHTLCStatus::Forward(PendingHTLCInfo { ref routing, .. }) => {
+				match routing {
+					PendingHTLCRouting::Receive { payment_data, keysend_preimage: Some(preimage), .. } => {
+						let hash = PaymentHash(Sha256::hash(&preimage.0[..]).into_inner());
+						let default_invoice_expiry_secs = 60 * 60; // 1 hour
+						self.set_payment_hash_secret_map(hash, Some(preimage.clone()), Some(payment_data.payment_secret), Some(payment_data.total_msat), default_invoice_expiry_secs, 0).unwrap();
```

Comment:
Hmm, I don't want to have to puzzle through whether this is DoS-y or if there's some way a peer can cause us to insert something into this map and it never be cleaned up.